replay_server/utils/greatest_common_replay.py

```diff
@@ -0,0 +1,166 @@
+        buffers_positions: List[int],
+        position: int,
+        size: int =-1
+) -> Optional[bytes]:
+    """
+    That method will return greatest common stream from position, with some size.
+    """
+
+    greatest_common_stream: Optional[bytes] = None
+    greatest_common_count = 0
+
+    if len(buffers) <= 2:
+        buffer = buffers[0]
+        buffer_position = buffers_positions[0]
+
+        # if second stream is bigger than first one, we get second one
+        if len(buffers) > 1 and get_buffer_size(buffers[1]) > get_buffer_size(buffers[0]):
+            buffer = buffers[1]
+            buffer_position = buffers_positions[1]
+
+        return _get_stream_part(buffer, buffer_position + position, size)
+
+    stream_information = get_common_buffers_info(buffers, buffers_positions, position, size)
+    stream_keys = stream_information.keys()
+
+    # choose which part has most common hits.
+    for file_no in stream_keys:
+        common_size = len(stream_information[file_no]["common_file_nos"])
+        if greatest_common_count < common_size:
+            greatest_common_stream = stream_information[file_no]["stream_part"]
+            greatest_common_count = common_size
+
+    return greatest_common_stream
+
+
+def get_common_buffers_info(
+        buffers: List[FileIO],
+        buffers_positions: List[int],
+        position: int,
+        size: int = -1
+) -> Dict:
+    """
+    Collects "common" data
+    """
+    # file handler id, file length, common stream matches
+    stream_information: Dict[int, Dict[str, int]] = {}
+
+    # get buffers information: file handler id and file length
+    for i, buffer in enumerate(buffers):
+        file_no = buffer.fileno()
+        file_size = get_buffer_size(buffer)
+        stream_part = _get_stream_part(buffer, buffers_positions[i] + position, size)
+        stream_information[file_no] = {"file_size": file_size, "stream_part": stream_part}
+
+    # compare everybody to everybody
+    stream_file_nos = stream_information.keys()
+    for file_no1 in stream_file_nos:
+        for file_no2 in stream_file_nos:
+            stream_information[file_no1].setdefault("common_file_nos", [])
+            if file_no1 == file_no2:
+                continue
+            stream_part1 = stream_information[file_no1]["stream_part"]
+            stream_part2 = stream_information[file_no2]["stream_part"]
+            if stream_part2 in stream_part1:
+                # register common hit count
+                stream_information[file_no1]["common_file_nos"].append(file_no2)
+
+    return stream_information
+
+
+def get_replay(paths: List[str]) -> str:
+    """
+    Problem: during the game some people can loose connectivity between them, so somebody being dropped out.
+    But all of them are still streaming replays. So there are multiple "truth" created.
+
+    Method will return sorted list of common groups, with biggest common replay.
+    That biggest one should be saved and streamed.
```

Comment:
This method is used just for saving the "most common" replay, right? I don't recall seeing it anywhere in reader code that does the merging.